framework/midi/miditypes.h

```diff
@@ -145,31 +145,109 @@ struct Event {
+        str += "channel: " + std::to_string(m_channel);
+        str += ", type: " + type_to_string(m_type);
+        switch (m_type) {
         case EventType::ME_NOTEON: {
-            str += ", key: " + std::to_string(a);
-            str += ", vel: " + std::to_string(b);
+            str += ", key: " + std::to_string(note());
+            str += ", vel: " + std::to_string(velocity());
         } break;
         case EventType::ME_NOTEOFF: {
-            str += ", key: " + std::to_string(a);
+            str += ", key: " + std::to_string(note());
         } break;
         case EventType::ME_CONTROLLER: {
-            str += ", cc: " + cc_to_string(a);
-            str += ", val: " + std::to_string(b);
+            str += ", cc: " + cc_to_string(controller());
+            str += ", val: " + std::to_string(value());
         } break;
         case EventType::ME_PITCHBEND: {
-            int pitch = b << 7 | a;
-            str += ", pitch: " + std::to_string(pitch);
+            str += ", pitch: " + std::to_string(pitch());
         } break;
         default:
-            str += ", a: " + std::to_string(a);
-            str += ", b: " + std::to_string(b);
+            str += "{";
+            for (auto& d : m_data) {
+                str += " " + std::to_string(d);
+            }
+            str += "}";
         }
 
         return str;
     }
+
+    channel_t channel() const { return m_channel; }
+    void setChannel(channel_t channel) { m_channel = channel; }
+
+    EventType type() const { return m_type; }
+    void setType(EventType type) { m_type = type; }
+
+    uint8_t note() const
+    {
+        assertType({ EventType::ME_NOTEON, EventType::ME_NOTEOFF });
+        return m_data[0];
+    }
+
+    void setNote(uint8_t value)
+    {
+        assertType({ EventType::ME_NOTEON, EventType::ME_NOTEOFF });
+        m_data[0] = value;
+    }
+
+    uint8_t velocity() const
+    {
+        assertType({ EventType::ME_NOTEON });
+        return m_data[1];
+    }
+
+    void setVelocity(uint8_t value)
+    {
+        assertType({ EventType::ME_NOTEON });
+        m_data[1] = value;
+    }
+
+    uint8_t pitch() const
+    {
+        assertType({ EventType::ME_PITCHBEND });
+        return m_data[1] << 7 | m_data[0];
+    }
+
+    void setPitch(uint8_t value)
+    {
+        m_data[0] = value & 0x7F;
+        m_data[1] = value >> 7;
+    }
+
+    uint8_t controller() const
+    {
+        assertType({ EventType::ME_CONTROLLER });
+        return m_data[0];
+    }
+
+    void setController(uint8_t value)
+    {
+        assertType({ EventType::ME_CONTROLLER });
+        m_data[0] = value;
+    }
+
+    uint8_t value() const
+    {
+        assertType({ EventType::ME_CONTROLLER, EventType::ME_PROGRAM });
+        return type() == EventType::ME_CONTROLLER ? m_data[1] : m_data[0];
+    }
+
+    void setValue(uint8_t value)
+    {
+        assertType({ EventType::ME_CONTROLLER, EventType::ME_PROGRAM });
+        m_data[1] = value;

```

Comment:
Probably should match  `type() == EventType::ME_CONTROLLER ? m_data[1] : m_data[0];`

Reply:
fixed